sycl/include/sycl/detail/kernel_launch_helper.hpp

```diff
@@ -251,20 +254,325 @@ struct KernelWrapper<
   }
 }; // KernelWrapper struct
 
+// This struct is inherited by handler_impl and sycl::handler is using the
+// static methods.
 struct KernelLaunchPropertyWrapper {
-  template <typename KernelName, typename PropertyProcessor,
-            typename KernelType>
-  static void parseProperties([[maybe_unused]] PropertyProcessor h,
-                              [[maybe_unused]] const KernelType &KernelFunc) {
+  // Changing values in this will break ABI/API.
+  enum class StableKernelCacheConfig : int32_t {
+    Default = 0,
+    LargeSLM = 1,
+    LargeData = 2
+  };
+
+  // This struct is used to store kernel launch properties.
+  // std::optional is used to indicate that the property is not set.
+  // In some code paths, kernel launch properties are set multiple times
+  // for the same kernel, i.e why using std::optional to avoid overriding
+  // previously set properties.
+  struct KernelLaunchPropertiesT {
+    std::optional<ur_kernel_cache_config_t> MKernelCacheConfig;
+    std::optional<bool> MKernelIsCooperative;
+    std::optional<uint32_t> MKernelWorkGroupMemorySize;
+    std::optional<bool> MKernelUsesClusterLaunch;
+    size_t MKernelClusterDims;
+    std::array<size_t, 3> MKernelClusterSize;
+
+    KernelLaunchPropertiesT()
+        : MKernelCacheConfig(std::nullopt), MKernelIsCooperative(std::nullopt),
+          MKernelWorkGroupMemorySize(std::nullopt),
+          MKernelUsesClusterLaunch(std::nullopt), MKernelClusterDims(0),
+          MKernelClusterSize{0, 0, 0} {}
+
+    KernelLaunchPropertiesT &operator=(const KernelLaunchPropertiesT &Other) {
+      if (Other.MKernelCacheConfig)
+        MKernelCacheConfig = Other.MKernelCacheConfig;
+
+      if (Other.MKernelIsCooperative)
+        MKernelIsCooperative = Other.MKernelIsCooperative;
+
+      if (Other.MKernelWorkGroupMemorySize)
+        MKernelWorkGroupMemorySize = Other.MKernelWorkGroupMemorySize;
+
+      if (Other.MKernelUsesClusterLaunch) {
+        MKernelUsesClusterLaunch = Other.MKernelUsesClusterLaunch;
+        MKernelClusterDims = Other.MKernelClusterDims;
+        MKernelClusterSize = Other.MKernelClusterSize;
+      }
+
+      return *this;
+    }
+
+    KernelLaunchPropertiesT(const KernelLaunchPropertiesT &Other) = default;
+    KernelLaunchPropertiesT(KernelLaunchPropertiesT &&Other) = default;
+    ~KernelLaunchPropertiesT() = default;
+  };
+
+  KernelLaunchPropertiesT KLProps;
+
+  // Set default values for kernel launch properties.
+  KernelLaunchPropertyWrapper() {
+    KLProps.MKernelCacheConfig = UR_KERNEL_CACHE_CONFIG_DEFAULT;
+    KLProps.MKernelIsCooperative = false;
+    KLProps.MKernelWorkGroupMemorySize = 0;
+    KLProps.MKernelUsesClusterLaunch = false;
+    KLProps.MKernelClusterDims = 0;
+    KLProps.MKernelClusterSize.fill(0);
```

Comment:
`MKernel*` made sense when it was part of `handler`, now that the class itself is specific for the kernel properties, you can name properties without "Kernel" in them, e.g. `MIsCooperative`.